custom-assembly-report-generator/generate_custom_assembly_report.py

```diff
@@ -0,0 +1,184 @@
+    contiginfo_query = "select distinct * from ({0}) temp where contig_name not in ({1})".format(
+        " union all ".join(["select contig_name, coalesce(contig_chr, contig_name) as chromosome, "
+                            "group_term as assembly_unit, contig_length from dbsnp_{0}.{1}"
+                           .format(species_info["database_name"], contiginfo_table_name)
+                            for contiginfo_table_name in contiginfo_tables]),
+        ','.join(["'{0}'".format(elem) for elem in refseq_accessions_from_assembly_report])
+    )
+    pg_cursor.execute(contiginfo_query)
+    accessions = pg_cursor.fetchall()
+    pg_cursor.close()
+    pg_conn.close()
+    return accessions
+
+
+def download_assembly_report(assembly_accession):
+    assembly_report_url = get_assembly_report_url(assembly_accession)
+    assembly_report_path = os.getcwd() + os.path.sep + os.path.basename(assembly_report_url)
+    urllib.urlretrieve(assembly_report_url, assembly_report_path)
+    urllib.urlcleanup()
+    return assembly_report_path
+
+
+def insert_absent_genbank_accessions_in_assembly_report(species_db_info, assembly_report_dataframe,
+                                                        genbank_equivalents_dataframe):
+    absent_refseq_accessions = \
+        get_absent_refseq_accessions(species_db_info, set(assembly_report_dataframe["RefSeq-Accn"]))
+    if len(absent_refseq_accessions) == 0:
+        print("INFO: No GenBank accessions are absent in the assembly report for " + species_db_info["database_name"])
+    new_assembly_report_entry_index = len(assembly_report_dataframe)
+    for refseq_accession, chromosome, assembly_unit, sequence_length in absent_refseq_accessions:
+        equivalent_genbank_accession = \
+            genbank_equivalents_dataframe[genbank_equivalents_dataframe["rs_acc"] == refseq_accession]["gb_acc"]
+        if len(equivalent_genbank_accession) == 0:
+            print("WARN: Could not find equivalent GenBank accession for RefSeq accession: " + refseq_accession)
+        else:
+            equivalent_genbank_accession = equivalent_genbank_accession.values[0]
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "Sequence-Role"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "Assigned-Molecule"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "Assigned-Molecule-Location/Type"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "GenBank-Accn"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "Relationship"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "RefSeq-Accn"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "Assembly-Unit"], \
+            assembly_report_dataframe.loc[new_assembly_report_entry_index, "Sequence-Length"] \
+            = "assembled-molecule", \

```

Comment:
Are we sure we want to use the pair `assembled-molecule` and `Chromosome`? I think that would cause issues when a chromosome is found multiple times in the assembly report. I would prefer to use `scaffold` or similar.

Reply:
Agree. My understanding is that we are overwriting the contigs already present, but even then we could get collision with other assembled molecules that point to the same contig.